core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -19,16 +19,27 @@ use zksync_crypto::hasher::{keccak::KeccakHasher, Hasher};
 /// we unlikely to ever hit.
 const MAX_TREE_DEPTH: usize = 32;
 
-/// In-memory Merkle tree of bounded depth (no more than 10).
+/// In-memory Merkle tree of bounded depth (no more than 32).
 ///
 /// The tree is left-leaning, meaning that during its initialization, the size of a tree
 /// can be specified larger than the number of provided leaves. In this case, the remaining leaves
 /// will be considered to equal `[0_u8; LEAF_SIZE]`.
+///
+/// The tree has dynamic size, meaning that it can grow by a factor of 2 when the number of leaves
+/// exceeds the current tree size. It does not shrink.
+///
+/// The tree is optimized for the case when the queries are performed on the rightmost leaves
+/// and the leftmost leaves are cached (trimmed). Caching enables the merkle roots and paths to be computed
+/// in `O(max(n, depth))` time, where `n` is the number of uncached leaves (in contrast to the total number of
+/// leaves). Cache itself only takes up `O(depth)` space. However, caching prevents the retrieval of paths to the
+/// cached leaves.
 #[derive(Debug, Clone)]
 pub struct MiniMerkleTree<const LEAF_SIZE: usize, H = KeccakHasher> {
     hasher: H,
-    hashes: Box<[H256]>,
+    hashes: VecDeque<H256>,
     binary_tree_size: usize,
+    start_index: usize,
+    cache: Vec<Option<H256>>,
```

Comment:
Better to include what does `cache[i]` contains. At the start I thought that `cache[0]` is for root, but then when reading I realized that it is not